File: internal_testsuites/golang/testsuite/startosis_add_service_test/startosis_add_service_with_empty_ports_test.go
package startosis_add_service_test

import (
	"context"
	"github.com/kurtosis-tech/kurtosis-cli/golang_internal_testsuite/test_helpers"
	"github.com/sirupsen/logrus"
	"github.com/stretchr/testify/require"
	"testing"
)

const (
	addServiceWithEmptyPortsTestName = "add-service-empty-ports"
	isPartitioningEnabled            = false
	defaultDryRun                    = false
	emptyArgs                        = "{}"

	serviceId  = "docker-getting-started"
	serviceId2 = "docker-getting-started-2"

	starlarkScriptWithEmptyPorts = `
DOCKER_GETTING_STARTED_IMAGE = "docker/getting-started:latest"
SERVICE_ID = "` + serviceId + `"

def run(args):
	print("Adding service " + SERVICE_ID + ".")
	
	config = struct(
		image = DOCKER_GETTING_STARTED_IMAGE,
		ports = {}
	)
	
	add_service(service_id = SERVICE_ID, config = config)
	print("Service " + SERVICE_ID + " deployed successfully.")
`

	starlarkScriptWithoutPorts = `
DOCKER_GETTING_STARTED_IMAGE = "docker/getting-started:latest"
SERVICE_ID = "` + serviceId2 + `"

print("Adding service " + SERVICE_ID + ".")

config = struct(
    image = DOCKER_GETTING_STARTED_IMAGE,
)

add_service(service_id = SERVICE_ID, config = config)
print("Service " + SERVICE_ID + " deployed successfully.")
`
)

var serviceIds = []string{serviceId, serviceId2}
var starlarkScriptsToRun = []string{starlarkScriptWithEmptyPorts, starlarkScriptWithoutPorts}

func TestAddServiceWithEmptyPortsAndWithoutPorts(t *testing.T) {
	ctx := context.Background()

	// ------------------------------------- ENGINE SETUP ----------------------------------------------
	enclaveCtx, destroyEnclaveFunc, _, err := test_helpers.CreateEnclave(t, ctx, addServiceWithEmptyPortsTestName, isPartitioningEnabled)
	require.NoError(t, err, "An error occurred creating an enclave")
	defer destroyEnclaveFunc()

	// ------------------------------------- TEST RUN ----------------------------------------------

	for starlarkScripIndex, starlarkScript := range starlarkScriptsToRun {
		logrus.Infof("Executing Starlark script...")
		logrus.Debugf("Starlark script content: \n%v", starlarkScript)

		outputStream, _, err := enclaveCtx.RunStarlarkScript(ctx, starlarkScript, emptyArgs, defaultDryRun)
		require.NoError(t, err, "Unexpected error executing starlark script")
		scriptOutput, _, interpretationError, validationErrors, executionError := test_helpers.ReadStreamContentUntilClosed(outputStream)

		expectedScriptOutput := `Adding service ` + serviceIds[starlarkScripIndex] + `.
Service '` + serviceIds[starlarkScripIndex] + `' added with internal ID '[a-z-0-9]+'
Service ` + serviceIds[starlarkScripIndex] + ` deployed successfully.
`
		require.Nil(t, interpretationError, "Unexpected interpretation error.")
		require.Empty(t, validationErrors, "Unexpected validation error")
		require.Nil(t, executionError, "Unexpected execution error")
		require.Regexp(t, expectedScriptOutput, scriptOutput)
		logrus.Infof("Successfully ran Starlark script")

		// Ensure that the service is listed
		expectedNumberOfServices := starlarkScripIndex + 1
		serviceInfos, err := enclaveCtx.GetServices()
		require.Nil(t, err)
		actualNumberOfServices := len(serviceInfos)
		require.Equal(t, expectedNumberOfServices, actualNumberOfServices)
	}
}

File: internal_testsuites/typescript/src/testsuite/startosis_add_service_test/startosis_add_service_with_empty_ports.test.ts
import {createEnclave} from "../../test_helpers/enclave_setup";
import log from "loglevel";
import { Result } from "neverthrow"
import {ServiceGUID, ServiceID} from "kurtosis-sdk";

const ADD_SERVICE_WITH_EMPTY_PORTS_TEST_NAME = "add-service-empty-ports"
const IS_PARTITIONING_ENABLED = false
const DEFAULT_DRY_RUN = false
const EMPTY_ARGS = "{}"
const SERVICE_ID = "docker-getting-started"
const SERVICE_ID_2 = "docker-getting-started-2"

const STARLARK_SCRIPT_WITH_EMPTY_PORTS = `
DOCKER_GETTING_STARTED_IMAGE = "docker/getting-started:latest"
SERVICE_ID = "${SERVICE_ID}"

def run(plan):
    plan.print("Adding service " + SERVICE_ID + ".")
    
    config = ServiceConfig(
        image = DOCKER_GETTING_STARTED_IMAGE,
        ports = {}
    )
    
    plan.add_service(service_id = SERVICE_ID, config = config)
    plan.print("Service " + SERVICE_ID + " deployed successfully.")`

const STARLARK_SCRIPT_WITHOUT_PORTS = `
DOCKER_GETTING_STARTED_IMAGE = "docker/getting-started:latest"
SERVICE_ID = "${SERVICE_ID_2}"

def run(plan, args):
    plan.print("Adding service " + SERVICE_ID + ".")
    
    config = ServiceConfig(
        image = DOCKER_GETTING_STARTED_IMAGE,
    )
    
    plan.add_service(service_id = SERVICE_ID, config = config)
    plan.print("Service " + SERVICE_ID + " deployed successfully.")`


jest.setTimeout(180000)

test("Test add service with empty and without ports test", TestAddServiceWithEmptyAndWithoutPorts)

async function TestAddServiceWithEmptyAndWithoutPorts() {

    const serviceIds: Array<string> = new Array<string>()
    serviceIds.push(SERVICE_ID)
    serviceIds.push(SERVICE_ID_2)

    const starlarkScriptsToRun: Array<string> = new Array<string>()
    starlarkScriptsToRun.push(STARLARK_SCRIPT_WITH_EMPTY_PORTS)
    starlarkScriptsToRun.push(STARLARK_SCRIPT_WITHOUT_PORTS)

    // ------------------------------------- ENGINE SETUP ----------------------------------------------
    const createEnclaveResult = await createEnclave(ADD_SERVICE_WITH_EMPTY_PORTS_TEST_NAME, IS_PARTITIONING_ENABLED)

    if(createEnclaveResult.isErr()) { throw createEnclaveResult.error }

    const { enclaveContext, stopEnclaveFunction } = createEnclaveResult.value

    try {
        // ------------------------------------- TEST SETUP ----------------------------------------------

        for (let i=0; i < starlarkScriptsToRun.length; i++) {
            const starlarkScript:string = starlarkScriptsToRun[i]
            const serviceId:string = serviceIds[i]
            log.info("Executing Starlark script...");
            log.debug(`Starlark script content: \n%v ${starlarkScript}`);
            const runResult = await enclaveContext.runStarlarkScriptBlocking(starlarkScript, EMPTY_ARGS, DEFAULT_DRY_RUN)
            if (runResult.isErr()) {
                log.error("Unexpected error executing Starlark script");
                throw runResult.error;
            }

            const expectedScriptOutputRegexpPattern = `Adding service ${serviceId}.
Service '${serviceId}' added with service GUID '[a-z-0-9]+'
Service ${serviceId} deployed successfully.
`;
            const expectedScriptOutputRegexp = new RegExp(expectedScriptOutputRegexpPattern)

            expect(runResult.value.interpretationError).toBeUndefined();
            expect(runResult.value.validationErrors).toEqual([]);
            expect(runResult.value.executionError).toBeUndefined();
            expect(runResult.value.runOutput).toMatch(expectedScriptOutputRegexp);
            log.info("Successfully ran Starlark script");

            // ------------------------------------- TEST RUN ----------------------------------------------

            // Ensure that the service is listed
            const expectedNumberOfServices: number = i + 1;
            const getServiceIdsPromise: Promise<Result<Map<ServiceID, ServiceGUID>, Error>> = enclaveContext.getServices();
            const getServiceIdsResult = await getServiceIdsPromise;
            if(getServiceIdsResult.isErr()) {
                log.error(`An error occurred getting service IDs`);
                throw getServiceIdsResult.error;
            }

            const serviceIdMap: Map<ServiceID, ServiceGUID> = getServiceIdsResult.value;

            const actualNumberOfServices: number = serviceIdMap.size

            if (expectedNumberOfServices !== actualNumberOfServices) {
                throw new Error(`Expected to receive ${expectedNumberOfServices} services from get services, but ${actualNumberOfServices} were received`)
            }
        }
    }
    finally {
        stopEnclaveFunction()
    }

    jest.clearAllTimers()
}
